### PR TITLE
Add most-likely-miss email notification (second attempt)

### PR DESCRIPTION
**Pull request by Mykhailenko**

### CODE REVIEW

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 1]
```diff
@@ -65,23 +65,39 @@
...
     private static final Set<JobId> alreadyNotified = new HashSet<>();
 
-    private synchronized void logIfJobWillMissItsDeadline(List<JobDescriptor> jobs) {
+    private synchronized void logMostLikelyMissedJobs(List<JobDescriptor> jobs) {
         Date now = new Date();
         jobs.stream()
             .map(jobDescriptor -> ((JobDescriptorImpl) jobDescriptor).getInternal())
             .filter(job -> job.getJobDeadline().isPresent())
-            .filter(job -> job.getJobInfo().getStartTime() >= 0)
+            .filter(job -> job.getJobInfo().getStartTime() < 0)
```
Comment (vinseon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: Can you please explain the logic behind in a comment?

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        try {
+            new SendMail().sender(getTo(), getSubject(), getBody());
+            return true;
+        } catch (EmailException | IOException e) {
+            throw new JobEmailNotificationException("Error sending email: " + e.getMessage(), e);
+        }
+    }
+
+    public void checkAndSend() {
```
Comment (vinseon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java]: Most likely `sendAndCheck` ;-)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 3]
```diff
@@ -65,23 +65,39 @@
...
             .filter(job -> {
                 Date deadline = getEffectiveDeadline(job);
                 Date expeFinish = getEffectiveExpectedExecutionTime(job, now);
                 return deadline.compareTo(expeFinish) < 0;
             })
             .forEach(job -> {
                 if (!alreadyNotified.contains(job.getId())) {
-                    LOGGER.warn(String.format("Job[id=%s] might miss its deadline (expected finish: %s after deadline: %s)",
+                    final Date expectedFinishingTime = getEffectiveExpectedExecutionTime(job, now);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "expectedFinishingTime" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 4]
```diff
@@ -65,23 +65,39 @@
...
+                    final Date effectiveDeadline = getEffectiveDeadline(job);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "effectiveDeadline" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 5]
```diff
@@ -65,23 +65,39 @@
 
     private static final Set<JobId> alreadyNotified = new HashSet<>();
 
-    private synchronized void logIfJobWillMissItsDeadline(List<JobDescriptor> jobs) {
+    private synchronized void logMostLikelyMissedJobs(List<JobDescriptor> jobs) {
         Date now = new Date();
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MAJOR Remove this useless assignment to local variable "now". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1854)
MINOR Remove this unused "now" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 6]
```diff
@@ -65,23 +65,39 @@
...
+                    final Date expectedFinishingTime = getEffectiveExpectedExecutionTime(job, now);
+                    final Date effectiveDeadline = getEffectiveDeadline(job);
+                    LOGGER.warn(String.format("Job[id=%s] will most likely miss its deadline (expected finish: %s, after deadline: %s)",
                                               job.getId().value(),
-                                              getEffectiveExpectedExecutionTime(job, now),
-                                              getEffectiveDeadline(job)));
+                                              expectedFinishingTime,
+                                              effectiveDeadline));
+
+                    JobMostLikelyMissedEmailNotification notificationSender;
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "notificationSender" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 7]
```diff
@@ -65,23 +65,39 @@
...
-    private synchronized void logIfJobWillMissItsDeadline(List<JobDescriptor> jobs) {
+    private synchronized void logMostLikelyMissedJobs(List<JobDescriptor> jobs) {
         Date now = new Date();
         jobs.stream()
             .map(jobDescriptor -> ((JobDescriptorImpl) jobDescriptor).getInternal())
             .filter(job -> job.getJobDeadline().isPresent())
-            .filter(job -> job.getJobInfo().getStartTime() >= 0)
+            .filter(job -> job.getJobInfo().getStartTime() < 0)
             .filter(job -> {
                 Date deadline = getEffectiveDeadline(job);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "deadline" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 8]
```diff
@@ -65,23 +65,39 @@
...
                 Date expeFinish = getEffectiveExpectedExecutionTime(job, now);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "expeFinish" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        return to;
+    }
+
+    private String getSubject() {
+        return String.format("WARNING: ProActive Job %s most likely will miss its deadline",
+                             internalJob.getId().value());
+    }
+
+    private static final SimpleDateFormat jobExecutionTimeFormat;
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java]: MAJOR Make "jobExecutionTimeFormat" an instance variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2885)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    public JobMostLikelyMissedEmailNotification(InternalJob internalJob, Date expectedFinishingTime, Date deadline,
+            Duration jobExecutionTime) {
+        this.internalJob = internalJob;
+        this.expectedFinishingTime = expectedFinishingTime;
+        this.deadline = deadline;
+        this.jobExecutionTime = jobExecutionTime;
+    }
+
+    public static void main(String[] args) {
+        System.out.println(new Date().getTime());
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        System.out.println(System.currentTimeMillis());
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/JobMostLikelyMissedEmailNotification.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 12]
```diff
@@ -65,24 +65,38 @@
...
+                return deadline.compareTo(finishingTime) < 0;
             })
             .forEach(job -> {
                 if (!alreadyNotified.contains(job.getId())) {
-                    LOGGER.warn(String.format("Job[id=%s] might miss its deadline (expected finish: %s after deadline: %s)",
-                                              job.getId().value(),
-                                              getEffectiveExpectedExecutionTime(job, now),
-                                              getEffectiveDeadline(job)));
                     alreadyNotified.add(job.getId());
+                    final Date deadline = getDeadlineAsAbsoluteDate(job);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "deadline" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 13]
```diff
@@ -65,24 +65,38 @@
...
+                    final Date finishingTime = getFinishingTimeIfJobStartedNow(job, now);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "finishingTime" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 14]
```diff
@@ -65,24 +65,38 @@
...
+                    LOGGER.warn(String.format("Job[id=%s] will most likely miss its deadline (expected finish: %s is after deadline: %s)",
+                                              job.getId().value(),
+                                              finishingTime,
+                                              deadline));
+
+                    JobMostLikelyMissedEmailNotification notificationSender;
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "notificationSender" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 15]
```diff
@@ -65,24 +65,38 @@
...
         jobs.stream()
             .map(jobDescriptor -> ((JobDescriptorImpl) jobDescriptor).getInternal())
             .filter(job -> job.getJobDeadline().isPresent())
-            .filter(job -> job.getJobInfo().getStartTime() >= 0)
+            .filter(job -> !job.getJobInfo().isStarted())
             .filter(job -> {
-                Date deadline = getEffectiveDeadline(job);
-                Date expeFinish = getEffectiveExpectedExecutionTime(job, now);
-                return deadline.compareTo(expeFinish) < 0;
+                Date deadline = getDeadlineAsAbsoluteDate(job);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "deadline" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java [hunk 16]
```diff
@@ -65,24 +65,38 @@
...
+                Date finishingTime = getFinishingTimeIfJobStartedNow(job, now);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java]: MINOR Remove this unused "finishingTime" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Mykhailenko** [comment]
Retest this please

**Comment by Mykhailenko** [comment]
Retest this please

**Comment by jrochas** [comment]
retest this please

**Pull request merged**